pkg/manifests/external_dns.go

```diff
@@ -419,48 +449,47 @@ func newExternalDNSDeployment(conf *config.Config, externalDnsConfig *ExternalDn
 	}
 }
 
-func generateResourceDeploymentArgs(rts ...ResourceType) []string {
+func generateResourceDeploymentArgs(rts ResourceTypes) []string {
 	var ret []string
-	for _, rt := range rts {
-		switch rt {
-		case ResourceTypeGateway:
-			ret = append(ret, []string{
-				"--source=gateway-httproute",
-				"--source=gateway-grpcroute",
-			}...)
-		case ResourceTypeIngress:
-			ret = append(ret, "--source=ingress")
-		}
+
+	if rts.Gateway {
+		ret = append(ret, []string{
+			"--source=gateway-httproute",
+			"--source=gateway-grpcroute",
+		}...)
+	}
+
+	if rts.Ingress {
+		ret = append(ret, "--source=ingress")
 	}
 
 	return ret
 }
 
-func addResourceSpecificRules(role *rbacv1.ClusterRole, resourceTypes ...ResourceType) {
-	for _, rt := range resourceTypes {
-		switch rt {
-		case ResourceTypeGateway:
-			role.Rules = append(role.Rules,
-				[]rbacv1.PolicyRule{
-					{
-						APIGroups: []string{""},
-						Resources: []string{"namespaces"},
-						Verbs:     []string{"get", "watch", "list"},
-					},
-					{
-						APIGroups: []string{"gateway.networking.k8s.io"},
-						Resources: []string{"gateways", "httproutes", "grpcroutes"},
-						Verbs:     []string{"get", "watch", "list"},
-					},
-				}...,
-			)
-		default:
-			role.Rules = append(role.Rules,
-				rbacv1.PolicyRule{
-					APIGroups: []string{"extensions", "networking.k8s.io"},
-					Resources: []string{"ingresses"},
+func addResourceSpecificRules(role *rbacv1.ClusterRole, resourceTypes ResourceTypes) {
```

Comment:
I think all these kinds of functions should be methods on the resourceTypes method and the resourceType should be an interface.

Reply:
same for generateResourceDeploymentArgs

Reply:
done